File: compiler/bench/mutrec4.c
// expected: 2000000
#include <stdio.h>
long g4(long n);
long h4(long n);
long i4(long n);
long f4(long n) {
    if (n == 0) return 0;
    return g4(n - 1) + 1;
}
long g4(long n) {
    if (n == 0) return 0;
    return h4(n - 1) + 1;
}
long h4(long n) {
    if (n == 0) return 0;
    return i4(n - 1) + 1;
}
long i4(long n) {
    if (n == 0) return 0;
    return f4(n - 1) + 1;
}
int main() {
    long sum = 0;
    for (long i = 0; i < 20000; i++) {
        sum += f4(100);
    }
    printf("%ld\n", sum);
    return 0;
}

File: compiler/bench/mutrec4.fs
\ expected: 2000000
\ 4 mutually recursive functions, 2M total calls
\ f->g->h->i->f chain

: f4 ( n -- result ) recursive
  dup 0= if exit then
  1- g4 1+ ;

: g4 ( n -- result ) recursive
  dup 0= if exit then
  1- h4 1+ ;

: h4 ( n -- result ) recursive
  dup 0= if exit then
  1- i4 1+ ;

: i4 ( n -- result ) recursive
  dup 0= if exit then
  1- f4 1+ ;

: main
  0 20000 0 do
    100 f4 +
  loop . cr ;
